hera_pspec/uvwindow.py

```diff
@@ -0,0 +1,1320 @@
+        little_h : boolean, optional
+                Whether to have cosmological length units be h^-1 Mpc or Mpc
+                Default: True (h^-1 Mpc).
+        x_orientation: str, optional
+            Orientation in cardinal direction east or north of X dipole.
+            Default keeps polarization in X and Y basis.
+            Used to convert pol str to num and convertly.
+        verbose : bool, optional
+            If True, print progress, warnings and debugging info to stdout.
+
+        """
+        # Summary attributes
+
+        # cosmology
+        if cosmo is None:
+            cosmo = conversions.Cosmo_Conversions()
+        self.cosmo = cosmo
+
+        # units
+        try:
+            bool(little_h)
+        except ValueError:
+            raise ValueError("little_h must be boolean")
+        self.little_h = bool(little_h)
+        if self.little_h:
+            self.kunits = units.h / units.Mpc
+        else:
+            self.kunits = units.Mpc**(-1)
+
+        # verbose
+        try:
+            bool(verbose)
+        except ValueError:
+            raise ValueError("verbose must be boolean")
+        self.verbose = bool(verbose)
+
+        # taper
+        try:
+            dspec.gen_window(taper, 1)
+        except ValueError:
+            raise ValueError("Wrong taper. See uvtools.dspec.gen_window"
+                             " for options.")
+        self.taper = taper
+
+        # polarisation
+        if isinstance(pols, str):
+            # if unique pol in string format (e.g. 'xx')
+            assert pols in ['pI', 'pQ', 'pV', 'pU', 'xx', 'yy', 'xy', 'yx'],\
+                "Wrong polarisation"
+            pols = (pols, pols)
+        elif isinstance(pols, int):
+            # if pol in integer format
+            if len(str(pols)) <= 2:
+                # if unique polarisation integer (e.g. -5)
+                assert pols in [1, 2, 4, 3, -5, -6, -7, -8], "Wrong polarisation"
+                # convert pol number to str according to AIPS Memo 117.
+                pols = (uvputils.polnum2str(pols, x_orientation=x_orientation),
+                        uvputils.polnum2str(pols, x_orientation=x_orientation))      
+            else:
+                # if polarisation pair integer
+                pols = uvputils.polpair_int2tuple(pols, pol_strings=True)
+        # if tuple of polarisation strings
+        elif isinstance(pols[0], str):
+            assert pols[0] and pols[1] in ['pI', 'pQ', 'pV', 'pU', 'xx', 'yy', 'xy', 'yx'],\
+                "Wrong polarisation"
+        # if tuple of polarisation integers
+        elif isinstance(pols[0], int):
+            assert pols[0] and pols[1] in [1, 2, 4, 3, -5, -6, -7, -8],\
+                "Wrong polarisation"
+            pols = (uvutils.polnum2str(pols[0], x_orientation=x_orientation),
+                    uvutils.polnum2str(pols[0], x_orientation=x_orientation))
+        else:
+            raise TypeError("Must feed pols as (tuple of) str or int.")
```

Comment:
feels like a LOT of code repetition here... maybe just convert to tuple first, then iteratively test each element.